pyro/infer/traceenum_elbo.py

```diff
@@ -261,6 +295,21 @@ def loss_and_grads(self, model, guide, *args, **kwargs):
         warn_if_nan(loss, "loss")
         return loss
 
+    def compute_marginals(self, model, guide, *args, **kwargs):
+        """
+        Computes marginal distributions at each model-enumerated sample site.
```

Comment:
Do we want to default to computing marginals for all enumeration sites, or only a single user-specified site at a time?  We could also compute and cache all marginals and only return one, since requesting them in a different order could reduce sharing.

Reply:
How about we pass in a list of requested sites, or say a `query_fn(msg)` as we do with `poutine.block`?

Reply:
> list of requested sites

This seems reasonable for now, since it's what `EmpiricalMarginal` currently takes as a query as well.  The computational question of whether to compute all marginals or only one seems more immediate.

Reply:
Ok, it appears to be pretty clean and cheap to compute and return marginals for all sites; it should be only as expensive as O(1) SVI steps, since we're `sharing_intermediates()` across all sites. How about we implement the all-sites version in this PR, and if in a future application we find computational burden, we can add a `kwarg` like `sites` or `_sites`?